src/number_theory/divisors.lean

```diff
@@ -35,29 +36,106 @@ open finset
 namespace nat
 variable (n : ℕ)
 
+/-- (Implementation) An auxiliary definition for `nat.divisors_antidiagonal`. -/
+private def divisors_antidiagonal' : finset (ℕ × ℕ) :=
+(Icc 1 n ×ˢ Icc 1 n).filter (λ x, x.fst * x.snd = n)
+
+lemma mem_divisors_antidiagonal' {x : ℕ × ℕ} :
+  x ∈ divisors_antidiagonal' n ↔ x.fst * x.snd = n ∧ n ≠ 0 :=
+begin
+  simp only [divisors_antidiagonal', finset.mem_Icc, ne.def, finset.mem_filter, finset.mem_product],
+  rw and_comm,
+  apply and_congr_right,
+  rintro rfl,
+  split; intro h,
+  { contrapose! h, simp [h, succ_le_iff], },
+  { rw [mul_eq_zero, decidable.not_or_iff_and_not] at h,
+    simp only [succ_le_of_lt (nat.pos_of_ne_zero h.1), succ_le_of_lt (nat.pos_of_ne_zero h.2),
+               true_and],
+    exact ⟨le_mul_of_pos_right (nat.pos_of_ne_zero h.2),
+      le_mul_of_pos_left (nat.pos_of_ne_zero h.1)⟩ }
+end
+
+open_locale pointwise
+
+/-- `divisors_antidiagonal n` is the `finset` of pairs `(x,y)` such that `x * y = n`.
+  As a special case, `divisors_antidiagonal 0 = ∅`. -/
+def divisors_antidiagonal : ℕ →* finset (ℕ × ℕ) :=
+{ to_fun := divisors_antidiagonal',
+  map_one' := by { ext ⟨i, j⟩, simp [mem_divisors_antidiagonal', mul_eq_one_iff] },
+  map_mul' := λ m n,
+  begin
+    rcases m.eq_zero_or_pos with rfl | hm,
+    { simp [divisors_antidiagonal', empty_product] },
+    rcases n.eq_zero_or_pos with rfl | hn,
+    { simp [divisors_antidiagonal', empty_product] },
+    ext ⟨i, j⟩,
+    simp only [mem_divisors_antidiagonal', mul_eq_zero, hm.ne', hn.ne', not_false_iff, ne.def,
+      and_true, or_self, mem_mul, prod.exists, prod.mk_mul_mk, prod.mk.inj_iff],
+    split,
+    { rintro h,
+      obtain ⟨⟨⟨i₁, hi₁⟩, i₂, hi₂⟩, rfl⟩ := prod_dvd_and_dvd_of_dvd_prod ⟨_, h⟩,
+      refine ⟨i₁, i₂, i / i₁, j / i₂, rfl, _, nat.mul_div_cancel' hi₁, nat.mul_div_cancel' hi₂⟩,
+      rw [div_mul_div_comm hi₁ hi₂, h],
+      exact nat.mul_div_cancel_left _ hm },
+    rintro ⟨_, _, _, _, rfl, rfl, rfl, rfl⟩,
+    exact mul_mul_mul_comm _ _ _ _
+  end }
+
 /-- `divisors n` is the `finset` of divisors of `n`. As a special case, `divisors 0 = ∅`. -/
-def divisors : finset ℕ := finset.filter (λ x : ℕ, x ∣ n) (finset.Ico 1 (n + 1))
+def divisors : ℕ →* finset ℕ := (monoid_hom.fst ℕ ℕ).image.comp divisors_antidiagonal
```

Comment:
This seems silly to me, the new definition is O(n^2) while the old was O(n)

Reply:
I don't think either of these definitions are intended to be evaluated directly, and even if you do want to evaluate them the lemma `divisors_apply` lets you convert to the O(n) version. Do you have a concrete suggestion to redefine this as a monoid hom in some other way, which is less "silly" to you?

Reply:
Let's get the dependent PR in first, then I'll have a go

Reply:
I guess if I'm after the really fast version, it would be:
```
#check nat.factors

def divisors_list : list ℕ :=
let sqn := nat.sqrt n in
if sqn*sqn = n then
  let l := list.filter (λ x : ℕ, x ∣ n) (list.range' 1 (sqn - 1)) in
  l ++ [sqn] ++ l.reverse.map ((/) n)
else
  let l := list.filter (λ x : ℕ, x ∣ n) (list.range' 1 sqn) in
  l ++ l.reverse.map ((/) n)

def divisors_fast : finset ℕ :=
⟨n.divisors_list, sorry⟩
```

Reply:
Right, this would be faster than both versions (though I suspect still not optimal!), but I see this as `norm_num`'s job rather than defeq's responsibility

Reply:
Can we compromise on defining `divisors_antidiagonal` in term of `divisors` and not the other way around? Either as
```lean
n.divisors.map ⟨λ d, (d, n/d), λ p₁ p₂, congr_arg prod.fst⟩ = n.divisors_antidiagonal
```
or
```lean
n.divisors.map ⟨λ d, (n/d, d), λ p₁ p₂, congr_arg prod.snd⟩ = n.divisors_antidiagonal
```